### ISSUE TITLE
Errors from -fsanitize=undefined

### ISSUE THREAD

**Issue by loveshack** [flame/blis#346]
FYI, if I build with the undefined behaviour sanitizer I see reports like this from make checkblis:

```
include/skx/blis.h:6566:29: runtime error: left shift of 7 by 29 places cannot be represented in type 'int'
include/skx/blis.h:6246:25: runtime error: left shift of 7 by 29 places cannot be represented in type 'int'
```
Configure with "CFLAGS=-fsanitize=undefined LDFLAGS=-lubsan" to reproduce (at least with gcc 8 or 9).

**Comment by fgvanzee** [comment]
@devinamatthews, can you help me understand what is happening here?

**Comment by devinamatthews** [comment]
Somewhere (aggregate blis.h so I don't know exactly where) there is a `x<<y` where `x=7`, `y=29`, and `x` is an `int`. That overflows which for signed types is undefined behavior. @loveshack can you send the context for those lines? Also, I think it would help @fgvanzee if you could provide a stack trace rather than him having to figure out UBSan.

**Comment by fgvanzee** [comment]
Based on the line numbers, it appears the issue is in the bitshifts that take place in the `bli_obj_comp_dt()` and `bli_obj_set_comp_dt()` static functions, which are defined in `frame/include/bli_obj_macro_defs.h`.

I think I understand the issue. the computation datatype spans bits 29 to 31 of the `objbits_t info` field of `obj_t`. That means the bitfield occupies the last bit (bit 31), which would be fine if it were `unsigned int`, but it less than fine for `signed int`. And it does appear I tried to head this off by declaring `objbits_t` as `uint32_t`. But turns out that's not enough. For example, in `bli_obj_set_comp_prec()`:
```c
static void bli_obj_set_comp_dt( num_t dt, obj_t* obj )
{
    obj->info = ( objbits_t )
                ( obj->info & ~BLIS_COMP_DT_BITS ) |
                ( dt << BLIS_COMP_DT_SHIFT );
}
```
the problem seems to be that `dt` is a `num_t`, which is `typedef'ed` as an `enum` and therefore signed. Is there any way to make an `enum` unsigned?

One obvious solution would be to move the computation datatype to the `info2` field, which is the "spillover" `objbits_t` I've placed in `obj_t` to handle additional bitfields beyond the original 32 bits.

Comments, @devinamatthews?

**Comment by devinamatthews** [comment]
Bitfields?

**Comment by devinamatthews** [comment]
Before you ask:

```C
typedef struct
{
    int dt : 3;
    int whatever : 5;
    ...
} info_t;

**Comment by fgvanzee** [comment]
Sorry, I still have to ask, because I've never seen that syntax.

By "bitfields", I was referring to the usage spelled out in `bli_type_defs.h`, which involves manually "declaring" certain bits for certain uses and using bitmasks and shifts and stuff to query and manipulate them.

**Comment by devinamatthews** [comment]
https://en.cppreference.com/w/c/language/bit_field

**Comment by fgvanzee** [comment]
Oh wow. If this is what I think it is, this will clean up a lot of fragile code.

**Comment by devinamatthews** [comment]
Welcome to 1989ish :).

**Comment by fgvanzee** [comment]
Is it safe to assume that by using the [bit field C language construct](https://en.cppreference.com/w/c/language/bit_field), the compiler ends up inserting the requisite shifts and masks as needed?

**Comment by devinamatthews** [comment]
Yes, what is doesn't guarantee is 1) padding, 2) order. There are standard ABIs that extend the language guarantee, though; e.g. ELF requires a particular layout so that program units compiled with different compilers are guaranteed to agree.

**Comment by fgvanzee** [comment]
Okay, so my interpretation of what you are saying is that switching to language-based bit fields may result in an increase in the size of an `obj_t`.

**Comment by devinamatthews** [comment]
If it does I don't think it will matter. But you can measure it pretty easily.

**Comment by loveshack** [comment]
> Based on the line numbers, it appears the issue is in the bitshifts that take place in the `bli_obj_comp_dt()` and `bli_obj_set_comp_dt()` static functions, which are defined in `frame/include/bli_obj_macro_defs.h`.

Right, according to the generated version in my source.  I've no
evidence this is actually causing a problem, but undefined behaviour can
be a problem with aggressive (-O3) optimization.

Some types are signed that I'd expect to be unsigned (counts and
dimensions?), but would be non-trivial to change and likely not worth
worrying about.  (GCC has sign-compare and sign-conversion warnings,
e.g. with -Wextra, that indicate some.)

I ran the address sanitizer without issue, by the way, and some time ago
I tried the thread sanitizer on an OpenMP build.  I tried ubsan now
hoping it would help with my ppc64 problem.
